### ISSUE TITLE
`in` operator.  (was: can Any::downcast_ref be public? after upgrading rhai it's private now)

### ISSUE THREAD

**Issue by GopherJ** [jonathandturner/rhai#124]

**Comment by schungx** [comment]
No reason why it can't.  I'll make it public.

Watch for it to show up at the latest PR https://github.com/jonathandturner/rhai/pull/123

**Comment by GopherJ** [comment]
Hello @schungx  cool, also I'm thinking if `rhai` can add `in` operator to detect if some element is in a list.

My use case is in casbin-rs: https://github.com/casbin/casbin-rs

We use rhai for evaluating string expression. Now rhai doesn't support in operator so we have to convert 

```
r.sub in ("alice", "bob")
```

into

```
inMatch(r.sub, ["alice", "bob"])
```

and register an `inMatch` function.

**Comment by schungx** [comment]
There is:

```
["alice", "bob"].contains(r.sub)
```

Read the `README`.  It has been significantly expanded.

**Comment by schungx** [comment]
You can probably write your own very simple function to do it.  Rhai doesn't provide one for arrays because, technically speaking, you can override the `==` operator and define your own equality comparisons.

```rust
fn contains(array, element) {
    for item in array {
        if item == element { return true; }
    }
    false
}
```

**Comment by schungx** [comment]
> My use case is in casbin-rs: https://github.com/casbin/casbin-rs

Your use case only seem to need expressions instead of full scripting.  Therefore, you should check out `compile_expression` which only compiles a subset of the language -- those to support expressions -- to an `AST`.  Then you can repeatedly evaluate that `AST` without needing to parse again.

You'd also want to exclude language features you don't need.  For example, you may not need the standard library of functions, so `no_stdlib`.  You may not need floating-point (`no_float`), or different integer types (`only_i32`).  And definitely you'd not need to define any functions (`no_function`).

**Comment by GopherJ** [comment]
@schungx that's actually not convinent, I'm using that method but I need to use regex to convert the following expressing:

```
r.sub in ("alice", "bob")
```

to 

```
contains(r.sub, ["alice", "bob"])
```

**Comment by GopherJ** [comment]
ideally we should support the javascript-like in operator in rhai

**Comment by schungx** [comment]
So what you're suggesting is:

```
Expression :    Expression  'in'  Expression (must evaluate to Array)
```

This is probably easily doable by adding an `in` infix operator.  Let me experiment a bit.

**Comment by schungx** [comment]
> contains(r.sub, ["alice", "bob"])

It might be easier for your code generation to convert it into:

```rust
(r.sub == "alice") || (r.sub == "bob")
```

Should be more efficient since arrays and strings do not need to be copied around.

**Comment by GopherJ** [comment]
It's already implemented for the conversion but I would like to have it supported in rhai

https://github.com/casbin/casbin-rs/blob/301420c7630cd3b6021f9738facf213e19080dc8/src/util.rs#L13

**Comment by schungx** [comment]
PR https://github.com/jonathandturner/rhai/pull/128 adds 'in' expressions.

**Comment by GopherJ** [comment]
Cool, could you notify me when it's merged and published? I'll test it. Thanks for your awesome work

**Comment by schungx** [comment]
It is already merged and published to `crates.io` as `0.11.1`.

**Comment by GopherJ** [comment]
@schungx  I mean #128

**Comment by schungx** [comment]
Yes, https://github.com/jonathandturner/rhai/pull/128 is published as `0.11.1`.

**Comment by GopherJ** [comment]
hello @schungx 

```
    #[test]
    fn test_in_operator() {
        let mut engine = Engine::new();

        assert!(engine.eval::<bool>("1 in [1, 2, 3]").unwrap());
        assert!(engine
            .eval::<bool>(r#""alice" in ["alice", "bob", "julien"]"#)
            .unwrap());
        assert!(engine
            .eval::<bool>(r#""unknow" in ["alice", "bob", "julien"]"#)
            .unwrap());
    }

```

I didn't succeed to pass this test, could you have a look
